File: remove_bg.py
from PIL import Image
import numpy as np
import sys

def remove_background(input_path, output_path):
    img = Image.open(input_path).convert("RGBA")
    data = np.array(img)
    
    # Get the background color from top-left pixel
    bg_color = data[0, 0, :3].astype(np.int32)
    
    # Calculate color distance and create alpha mask
    pixels = data[:, :, :3].astype(np.int32)
    dist = np.linalg.norm(pixels - bg_color, axis=2)
    
    # Soft thresholding for anti-aliased text
    # The background is uniform, so distance is near 0 for bg.
    # Text is dark, so distance is high.
    alpha = np.clip((dist - 15) * 5, 0, 255).astype(np.uint8)
    
    # Because there might be semi-transparent pixels, we should keep their original rgb
    # but for text where it's anti-aliased with background, we should ideally subtract background.
    # We will just invert the background with alpha channel.
    # To avoid white fringes on dark backgrounds:
    # Instead of original pixels, we can set the RGB to the darkest pixel color,
    # or just use the original RGB but with the new alpha mask.
    data[:, :, 3] = alpha
    
    Image.fromarray(data).save(output_path)
    print(f"Saved to {output_path}")

input_img = sys.argv[1]
output_img = sys.argv[2]
remove_background(input_img, output_img)

File: remove_bg.js
const Jimp = require('jimp');

async function removeBackground() {
    const inputPath = process.argv[2];
    const outputPath = process.argv[3];

    console.log("Reading image:", inputPath);
    const image = await Jimp.read(inputPath);

    // Top-left pixel is considered background
    const bgHex = image.getPixelColor(0, 0);
    const bgRGBA = Jimp.intToRGBA(bgHex);
    console.log("Background color detected:", bgRGBA);

    image.scan(0, 0, image.bitmap.width, image.bitmap.height, function (x, y, idx) {
        const r = this.bitmap.data[idx + 0];
        const g = this.bitmap.data[idx + 1];
        const b = this.bitmap.data[idx + 2];
        const a = this.bitmap.data[idx + 3];

        if (a === 0) return; // already transparent

        // Euclidean distance from background color
        const dist = Math.sqrt(Math.pow(r - bgRGBA.r, 2) + Math.pow(g - bgRGBA.g, 2) + Math.pow(b - bgRGBA.b, 2));

        // Soft threshold logic for anti-aliasing
        // If distance is less than 15, assume it's background and make fully transparent.
        // If distance is more than 60, assume it's foreground and keep opaque.
        // In between, scale alpha proportionally.
        let alpha = 255;
        const lowerThreshold = 15;
        const upperThreshold = 80;

        if (dist <= lowerThreshold) {
            alpha = 0;
        } else if (dist < upperThreshold) {
            alpha = ((dist - lowerThreshold) / (upperThreshold - lowerThreshold)) * 255;
        }

        // Apply alpha channel modification
        this.bitmap.data[idx + 3] = alpha;
    });

    await image.writeAsync(outputPath);
    console.log("Saved transparent image to", outputPath);
}

removeBackground().catch(err => {
    console.error(err);
    process.exit(1);
});
